src/core/core/network.cpp

```diff
@@ -344,35 +360,36 @@ namespace dsn
         _parser(std::move(parser)),
         _is_client(is_client),
         _matcher(_net.engine()->matcher()),
-        _message_count(0),
         _is_sending_next(false),
+        _message_count(0),
         _connect_state(is_client ? SS_DISCONNECTED : SS_CONNECTED),
         _message_sent(0),
         _delay_server_receive_ms(0)
     {
     }
 
-    bool rpc_session::on_disconnected(bool is_write)
+    bool rpc_session::on_disconnected(bool /*is_write*/)
     {
-        if (is_client())
+        if (set_disconnected())
         {
-            set_disconnected();
             rpc_session_ptr sp = this;
-            _net.on_client_session_disconnected(sp);
+            if (is_client())
+            {
+                _net.on_client_session_disconnected(sp);
+            }
+            else
+            {
+                _net.on_server_session_disconnected(sp);
+            }
+
+            clear_send_queue(false);
```

Comment:
@qinzuoyan this change introduces bugs - considering the case where messages are sending while the remote peer disconnected which causes the read operation callback calls this function first, which immediately causes invalid data access in write operations (e.g., using _sending_buffers). I'll come up the fix later.